### PR TITLE
Task CRUD

### PR DESCRIPTION
**Pull request by mla-wtag**
[JTJ-53](https://welltravel.atlassian.net/jira/software/projects/JTJ/boards/44?assignee=712020%3A0898e570-ccb9-4a9a-8fcd-6555abd40896&selectedIssue=JTJ-53)

**Description**
- The user table is linked to the task model through a many-to-many association using a junction table called "assignment."
- The admin role can view all tasks, while the user role can only view tasks assigned to them.
-  Best practice? : If I open the task that is assigned to two users, it will be accessed via the URL 'users/1%2F2/task'." (video: 0:38 seconds)

https://github.com/mla-wtag/ml_Journey/assets/141820916/b749e877-eddf-420f-814c-c89246756e19

### CODE REVIEW

File: config/locales/en.yml [hunk 1]
```diff
@@ -18,13 +18,20 @@ en:
     content: Content
     goals_today: Goals Today
     goals_tomorrow: Goals Tomorrow
+    description: Description
```
Comment (tgi-wtag) [inline config/locales/en.yml]: I belive you should run " i18n-tasks normalize" to sort alphabetically and if your on it you can also try "i18n-tasks health " i guess

**Review by chm-wtag — APPROVED** [review]

File: spec/factories/tasks.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+FactoryBot.define do
+  factory :task do
+    title { FFaker::Book.title }
+    description { FFaker::Lorem.paragraph }
+    date { FFaker::Time.between(1.year.ago, Date.today).to_date }
+    status { %w[todo progress done].sample }
```
Comment (evk-wtag) [inline spec/factories/tasks.rb]:
```suggestion
    status { Task.statuses.keys.sample }
```

File: app/models/ability.rb [hunk 3]
```diff
@@ -20,7 +20,7 @@ def admin_abilities
   end
 
   def user_abilities
-    can %i(create read update destroy confirm_email), [User, Journal]
+    can %i(create read update destroy confirm_email), [User, Journal, Task]
```
Comment (evk-wtag) [inline app/models/ability.rb:23]: Missing rspec

File: spec/models/assignment_spec.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+require 'rails_helper'
+RSpec.describe Assignment, type: :model do
+  describe 'associations' do
+    it { should belong_to(:user) }
```
Comment (evk-wtag) [inline spec/models/assignment_spec.rb:4]:
```suggestion
    it { is_expected.to belong_to(:user) }
```

File: spec/models/assignment_spec.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+    it { should belong_to(:task) }
```
Comment (evk-wtag) [inline spec/models/assignment_spec.rb:5]:
```suggestion
    it { is_expected.to belong_to(:user) }
    it { is_expected.to belong_to(:task) }
```

File: spec/models/users_spec.rb [hunk 6]
```diff
@@ -95,4 +95,10 @@
       expect(confirmed_user.confirmed?).to be_truthy
     end
   end
+
+  describe 'associations' do
+    it { should have_many(:assignments) }
```
Comment (evk-wtag) [inline spec/models/users_spec.rb]:
```suggestion
    it { is_expected.to have_many(:assignments) }
```
and so on...

File: db/schema.rb [hunk 7]
```diff
@@ -55,12 +64,13 @@
   end
 
   create_table "tasks", force: :cascade do |t|
-    t.string "task_name"
+    t.string "title"
```
Comment (evk-wtag) [inline db/schema.rb]: Should be null: false right? Since we have a model validation

File: spec/factories/tasks.rb (hunk 2 above)
Comment (mla-wtag) [reply spec/factories/tasks.rb]: should I also try to apply this suggestion on the users.rb factory ? 
_code snippet from the user factory:_ 
`role { %w[user_role admin_role].sample }`

Comment (evk-wtag) [reply spec/factories/tasks.rb]: Yes sure!

File: app/controllers/tasks_controller.rb [hunk 8]
```diff
@@ -1,2 +1,38 @@
 class TasksController < ApplicationController
+  load_and_authorize_resource :user
+  load_and_authorize_resource :task, through: :user
+
+  def create
```
Comment (evk-wtag) [inline app/controllers/tasks_controller.rb:5]: Missing controller spec?

File: spec/models/tasks_spec.rb [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      is_expected.to have_many(:assignments).dependent(:destroy)
+    end
+  end
+
+  describe 'Task association with assignment' do
+    it 'has many assignments' do
+      is_expected.to have_many(:assignments)
+    end
+    it { should have_many(:users).join_table(:assignments) }
+  end
```
Comment (tgi-wtag) [inline spec/models/tasks_spec.rb:62]: I believe this makes more sense.

```
describe 'Task association with assignment' do
    it 'has many assignments' do
      is_expected.to have_many(:assignments)
      is_expected.to have_many(:users).join_table(:assignments) 
    end
end
```

File: spec/controllers/tasks_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+      it 're-renders the :new template with unprocessable entity status' do
+        post :create, params: { user_id: user.id, task: invalid_attributes }
+        expect(response).to render_template(:new)
+        expect(response).to have_http_status(:unprocessable_entity)
+      end
+    end
+  end
+
+  describe 'PATCH #update' do
+    context 'with valid attributes' do
```
Comment (tgi-wtag) [inline spec/controllers/tasks_controller_spec.rb:52]: `before { patch :update, params: { user_id: user.id, id: task.id, task: { title: updated_title } } }`

if you add this line here you dont have to add it to the individual specs below, DRY your code :)

File: spec/controllers/tasks_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+      end
+
+      it 'redirects to the task show page' do
+        patch :update, params: { user_id: user.id, id: task.id, task: { title: 'Updated Title' } }
+        expect(response).to redirect_to user_task_path(user)
+      end
+    end
+
+    context 'with invalid attributes' do
+      it 'does not update the task' do
```
Comment (tgi-wtag) [inline spec/controllers/tasks_controller_spec.rb]: `before { patch :update, params: { user_id: user.id, id: task.id, task: { title: nil } } }`

same thing applies here DRY

File: spec/controllers/tasks_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+      it 're-renders the :edit template with unprocessable entity status' do
+        patch :update, params: { user_id: user.id, id: task.id, task: { title: nil } }
+        expect(response).to render_template(:edit)
+        expect(response).to have_http_status(:unprocessable_entity)
+      end
+    end
+  end
+
+  describe 'DELETE #destroy' do
+    let(:user1) { FactoryBot.create(:user) }
```
Comment (tgi-wtag) [inline spec/controllers/tasks_controller_spec.rb]: `before { delete :destroy, params: { user_id: user.id, id: task.id } }`

same here, DRY

File: spec/controllers/tasks_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    end
+
+    it 'renders the :index template' do
+      get :index, params: { user_id: user.id }
+      expect(response).to render_template :index
+    end
+  end
+
+  describe 'POST #create' do
+    context 'with valid attributes' do
```
Comment (tgi-wtag) [inline spec/controllers/tasks_controller_spec.rb:23]:
```
  describe 'POST #create' do
    context 'with valid attributes' do

      before { post :create, params: { user_id: user.id, task: task_attributes } }

      it 'creates a task' do
        expect(Task.count).to eq(1)
      end

      it 'redirects to the tasks index' do
        expect(response).to redirect_to user_tasks_path(user)
      end
    end
```

you can DRY your code with this befor...

File: spec/controllers/tasks_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+RSpec.describe TasksController, type: :controller do
+  let!(:user) { FactoryBot.create(:user) }
+  let!(:assignment) { FactoryBot.create(:assignment) }
+  let!(:task) { FactoryBot.create(:task, creator: user, user_ids: user.id) }
+  before { sign_in(user) }
+
+  describe 'GET #index' do
+    let(:task2) { FactoryBot.create(:task, creator: user, user_ids: user.id) }
+
```
Comment (tgi-wtag) [inline spec/controllers/tasks_controller_spec.rb]: ` before { get :index, params: { user_id: user.id } }`

DRY your code by adding this line here

File: spec/controllers/tasks_controller_spec.rb (hunk 13 above)
Comment (evk-wtag) [reply spec/controllers/tasks_controller_spec.rb:23]: Thanks @tgi-wtag for the suggestion!

In addition to this, I would suggest another way for making it look more organized ⬇️ 

```
  describe 'POST #create' do
    context 'with valid attributes' do
      let(:create_action) { post :create, params: { task: task_attributes } }

      it 'creates a task' do
        expect { create_action }.to change(Task, :count).by(1)
      end

      it 'redirects to the tasks index' do
        expect(create_action).to redirect_to user_tasks_path(user)
      end
    end
```

**Review by evk-wtag — APPROVED** [review]

**Review by lra-wtag — APPROVED** [review]

**Pull request merged**